src/ripple/app/misc/AMMHelpers.h

```diff
@@ -228,10 +230,62 @@ swapAssetIn(
     TIn const& assetIn,
     std::uint16_t tfee)
 {
-    return toAmount<TOut>(
-        getIssue(pool.out),
-        pool.out - (pool.in * pool.out) / (pool.in + assetIn * feeMult(tfee)),
-        Number::rounding_mode::downward);
+    if (auto const& rules = getCurrentTransactionRules();
+        rules && rules->enabled(fixAMMRounding))
+    {
+        // set rounding to always favor the amm. Clip to zero.
+        // calculate:
+        // pool.out -
+        // (pool.in * pool.out) / (pool.in + assetIn * feeMult(tfee)),
+        // and explicitly set the rounding modes
+        // Favoring the amm means we should:
+        // minimize:
+        // pool.out -
+        // (pool.in * pool.out) / (pool.in + assetIn * feeMult(tfee)),
+        // maximize:
+        // (pool.in * pool.out) / (pool.in + assetIn * feeMult(tfee)),
+        // (pool.in * pool.out)
+        // minimize:
+        // (pool.in + assetIn * feeMult(tfee)),
+        // minimize:
+        // assetIn * feeMult(tfee)
+        // feeMult is: (1-fee), fee is tfee/100000
+        // minimize:
+        // 1-fee
+        // maximize:
+        // fee
+        saveNumberRoundMode _{Number::getround()};
+
+        Number::setround(Number::upward);
+        auto const numerator = pool.in * pool.out;
+        auto const fee = getFee(tfee);
```

Comment:
Does it make sense to always round upwards `getFee()` and downward `feeMult()`. The fee is added to the pool so we want always to maximize it's effect. Can have the default just in case it needs to be changed in some instances:

    inline Number
    getFee(std::uint16_t tfee, Number::rounding_mode rounding = Number::rounding_mode::upward)
    {
        if (auto const& rules = getCurrentTransactionRules();
            rules && rules->enabled(fixAMMRounding))
        {
            saveNumberRoundMode rm(Number::setround(rounding));
            return Number{tfee} / AUCTION_SLOT_FEE_SCALE_FACTOR;
        }
        return Number{tfee} / AUCTION_SLOT_FEE_SCALE_FACTOR;
    }

    inline Number
    feeMult(std::uint16_t tfee, Number::rounding_mode rounding = Number::rounding_mode::downward)
    {
        if (auto const& rules = getCurrentTransactionRules();
            rules && rules->enabled(fixAMMRounding))
        {
            saveNumberRoundMode rm(Number::setround(rounding));
            return 1 - getFee(tfee);
        }
        return 1 - getFee(tfee);
    }